### ISSUE TITLE
Discogs: rank candidates before return in item_candidates

### ISSUE THREAD

**Issue by ghbrown** [beetbox/beets#4703]
### Proposed solution
Rank the candidates inside the `item_candidates` method of `discogs.py` before returning the top 10 of these results.

#### Goals
When matching singletons to Discogs, `item_candidates` currently returns a limited amount (10) of possibly bad matches (even if the query was ideal). A full list of candidates is currently constructed by iterating over the first 5 albums in the query response, and adding all contained tracks. The candidates returned are the first 10 elements of this list.

A trivial example on which this approach would fail is on the 11th song of an album. Even if the proper album is returned first in the query response, the user will never be able to match to the 11th song, since only the first 10 are returned.

#### Anti-goals
Ranking and sorting the candidates can be an expensive process. There may be as many as 500 candidate tracks whose similarity must be computed before sorting the array of similarities.

The features used to rank the results can also affect cost and quality of results. Do we use title/artist strings only, or essentially compare entire TrackInfo objects? The former would be faster but less reliable, while the latter would be more expensive and more reliable.

#### Notes
The original reason such a feature was not implemented was because it performed a pruning process on the candidates that the user had no control over, and 10 candidates was deemed to be sufficient in most cases via testing. However, having just used this feature for hours myself I can say that the current state is not ideal, and fails on all kinds of simple cases. However, ranking the candidates rather than pruning is a more cautious approach and will only increase the quality of matches over the current approach (and the pruning approach).

**Comment by JOJ0** [comment]
Thanks for addressing this shortcoming of the Discogs singleton feature. That is indeed a tricky one and I don't know which of your proposed ways would be the better one. Performance is always an issue already with beets....

I really don't know other than: Do you think it would be possible to get some draft implementations going and then compare which one peforms better or put differently: I trust in your decision on which way to go is better and am happy to review another pull request and help with testing it :-)

Hope that helps at least somehow.... :-/

**Comment by JOJ0** [comment]
> The features used to rank the results can also affect cost and quality of results. Do we use title/artist strings only, or essentially compare entire TrackInfo objects? The latter would be faster but less reliable, while the latter would be more expensive and more reliable.

I think you have a said latter twice her. I suggest you correct your original post since this sentence probably is a very important one describing the two ways you see how this could be implemented :-) HTH

**Comment by ghbrown** [comment]
> I think you have a typo here, you said "latter" twice here. I suggest you correct your original post

Corrected, thanks for catching this.

**Comment by JOJ0** [comment]
> The features used to rank the results can also affect cost and quality of results. Do we use title/artist strings only, or essentially compare entire TrackInfo objects? The former would be faster but less reliable, while the latter would be more expensive and more reliable.

After reading this again and using singleton search in my current task of fixing a very broken old mp3 collection I came to the conclusion that a more reliable approach should be preferred. Even though it might take longer to get results, in unattended imports that does not matter anyway.

**Comment by ghbrown** [comment]
@JOJ0 Just checking up on this. Are we still waiting for #4716 to be addressed?

**Comment by JOJ0** [comment]
Well, that is my bad. I wanted to finish that PR but got distracted by other beetsy and personal things.

It's finished except one thing (fixme link) but missing docs and changelog.

**Comment by JOJ0** [comment]
Hi @ghbrown I finally managed to finished the "singleton albums" feature. Note that I refactored it in this new PR: #4854 

I'd love to hear your opinion how it goes for your imports once you found time to use it a little "in production".

In the end I left the 10 candidates limit in-tact. We will see how that goes. We might want ot make that configurable if we find it sometimes is not enough.

Selecting the "correct" track from an album works like this now in item_candidates():

https://github.com/beetbox/beets/blob/c03bd26e3a80223633249e31907e08b4076cf514/beetsplug/discogs.py#L266-L274

This is the magic that selects the track using the string_dist method with an hardcoded distshreshold of 0.3. I stole this from the original PR FIXME and didn't invent this value myself. It seems to work quite well so far.

https://github.com/beetbox/beets/blob/c03bd26e3a80223633249e31907e08b4076cf514/beetsplug/discogs.py#L199-L231

**Comment by JOJ0** [comment]
Hi @ghbrown did you find the time  to use beets with this new feature merged? Do we still need a "ranking" feature? What do you think?

**Comment by ghbrown** [comment]
@JOJ0 Apologies for the insane delay on this. I used beets about a week ago (and for the first time in a long time) with the Discogs backend and the experience was much improved. I did not experience any issues resembling "query returned track but did not present it as an option" as I did before your merge, or if I did encounter this I never noticed.

Thanks for getting this working, it's a massive quality of life improvement for users.
